script/src/scheduler.rs

```diff
@@ -0,0 +1,875 @@
+                            exit_code_addr,
+                        } = state
+                        {
+                            if *target_vm_id == vm_id_to_run {
+                                joining_vms.push((*vm_id, *exit_code_addr));
+                            }
+                        }
+                    });
+                    // For all joining VMs, update exit code, then mark them as
+                    // runnable state.
+                    for (vm_id, exit_code_addr) in joining_vms {
+                        self.ensure_vms_instantiated(&[vm_id])?;
+                        let (_, machine) = self
+                            .instantiated
+                            .get_mut(&vm_id)
+                            .ok_or_else(|| Error::Unexpected("Unable to find VM Id".to_string()))?;
+                        machine
+                            .machine
+                            .memory_mut()
+                            .store8(&exit_code_addr, &u64::from_i8(code))?;
+                        machine.machine.set_register(A0, SUCCESS as u64);
+                        self.states.insert(vm_id, VmState::Runnable);
+                    }
+                    // Close fds
+                    self.fds.retain(|_, vm_id| *vm_id != vm_id_to_run);
+                    // Clear terminated VM states
+                    self.states.remove(&vm_id_to_run);
+                    self.instantiated.remove(&vm_id_to_run);
+                    self.suspended.remove(&vm_id_to_run);
+                }
+                self.total_cycles = self
+                    .total_cycles
+                    .checked_add(self.extra_cycles)
+                    .ok_or(Error::CyclesOverflow)?;
+                Ok(consumed_cycles + self.extra_cycles)
+            }
+            Err(Error::Yield) => {
+                self.total_cycles = self
+                    .total_cycles
+                    .checked_add(self.extra_cycles)
+                    .ok_or(Error::CyclesOverflow)?;
+                Ok(consumed_cycles + self.extra_cycles)
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    fn process_message_box(&mut self) -> Result<(), Error> {
+        let messages: Vec<Message> = self.message_box.lock().expect("lock").drain(..).collect();
+        for message in messages {
+            match message {
+                Message::Spawn(vm_id, args) => {
+                    // All fds must belong to the correct owner
+                    let mut fds_valid = true;
+                    for fd in &args.fds {
+                        if !(self.fds.contains_key(fd) && (self.fds[fd] == vm_id)) {
+                            self.ensure_vms_instantiated(&[vm_id])?;
+                            let (_, machine) =
+                                self.instantiated.get_mut(&vm_id).ok_or_else(|| {
+                                    Error::Unexpected("Unable to find VM Id".to_string())
+                                })?;
```

Comment:
`machine` are mutable borrows, using it as return values will make `self` also mutable borrows, writing codes like this in Rust is terrible.

Reply:
I don't think it's a problem in rust, please check the patch:
[ensure_vm_instantiated.diff.txt](https://github.com/nervosnetwork/ckb/files/15106702/ensure_vm_instantiated.diff.txt)

After we change all the snippets from `ensure_vms_instantiated` to `ensure_vm_instantiated`, the previous fn can be removed, and we can use a simpler method to resume from full state.

Reply:
This works in some places, but not in others, such as in `fn iterate()`.

```rs
let (context, machine) = self.ensure_vm_instantiated(&vm_id_to_run)?; // The function needs to be modified so that it returns context
context.set_base_cycles(self.total_cycles); // cannot use `self.total_cycles` because it was mutably borrowed use of borrowed `*self`
```